aten/src/ATen/core/type.cpp

```diff
@@ -753,6 +760,255 @@ TupleTypePtr TupleType::createNamed(
+bool NoneType::isSubtypeOfExt(const TypePtr& rhs, std::ostream *why_not) const {
+  if (rhs->kind() == OptionalType::Kind) {
+    return true;
+  }
+  if (auto rhs_ = rhs->cast<UnionType>()) {
+    return rhs_->can_hold_none();
+  }
+  return Type::isSubtypeOfExt(rhs, why_not);
+}
+
+// Remove nested Optionals/Unions during the instantiation of a Union or
+// an Optional. This populates `types` with all the types found during
+// flattening
+void flattenUnion(TypePtr& type, std::vector<TypePtr>& types) {
+  if (type->kind() == UnionType::Kind) {
+    for (auto inner_type : type->expect<UnionType>()->types()) {
+      flattenUnion(inner_type, types);
+    }
+  } else if (type->kind() == OptionalType::Kind) {
+    auto inner_type = type->expect<OptionalType>()->getElementType();
+    if (inner_type->kind() == UnionType::Kind
+        || inner_type->kind() == OptionalType::Kind) {
+      flattenUnion(inner_type, types);
+    } else {
+      types.emplace_back(inner_type);
+    }
+    types.emplace_back(NoneType::get());
+  } else {
+    types.emplace_back(type);
+  }
+}
+
+// Custom comparison for Union type auxiliary functions. We compare on
+// Kind (fast), then compare by the serialized representation
+// (slower, but guaranteed to break ties). We need to have the
+// TypeEqual struct for use in a `std::unordered_map` later.
+struct TypeEqual {
+  bool operator()(const TypePtr& a, const TypePtr& b) const {
+    return a->kind() == b->kind() && a->str() == b->str();;
+  }
+};
+
+// Helper function for `standardizeUnion`
+void filterDuplicateSubtypes(std::vector<TypePtr>& types) {
+  // If we call `unifyTypes(T, None)`, then `unifyTypes` returns
+  // `Optional[T]`. The `Optional` constructor calls the `Union`
+  // constructor, which calls `filterDuplicateSubtypes`...infinite
+  // recursion (until the stack explodes). So, let's just wrap
+  // `unifyTypes` to ensure that this condition doesn't happen. We
+  // only need to worry about this in the constructor
+  auto wrapper = [](const TypePtr t1, const TypePtr t2) -> c10::optional<TypePtr> {
+    // If we would have unified the types to an Optional
+    if ((t1 == NoneType::get()) ^ (t2 == NoneType::get())) {
+      return c10::nullopt;
+    }
+    return unifyTypes(t1, t2);
+  };
+
+  // Coalesce types and delete all duplicates. Moving from right to left
+  // through the vector, we try to unify the current element (`i`) with
+  // each element (`j`) before the "new" end of the vector (`end`).
+  // If we're able to unify the types at `types[i]` and `types[j]`, we
+  // decrement `end`, swap `types[j]` with the unified type, and
+  // break. Otherwise, we keep `end` where it is to signify that the
+  // new end of the vector hasn't shifted
+  size_t end_idx = types.size()-1;
+  for (size_t i = types.size()-1; i > 0; --i) {
+    for (size_t j = std::min(i-1, end_idx); ; --j) {
+      c10::optional<TypePtr> unified = wrapper(types[i], types[j]);
+      if (unified) {
+        types[j] = *unified;
```

Comment:
I don't think unifying is what you're after here - you just need to test for subtyping. Unification will find a common supertype even if the two args aren't directly in a subtype relationship, so in those cases you'll be merging rather than just deduping, which you don't want.

Like if A > B and A > C, and you have Union[B, C], with unification you'll wind up with Union[A], but that's a different type - you don't know what else A has in it besides B and C.

OTOH if you have Union[A, C], then you do want to wind up with Union[A], bc C is completely contained in A, so the two unions are equivalent in terms of the values they represent.

Reply:
cc @gmagogsfm @eellison 
I agree with Basil's point, and I'm changing the code accordingly. Basil and I talked offline, and we discussed that, right now, `unifyTypes` won't have the `A, B, C` problem because of how limited our type system is. So, this change is for forward compatibility more than anything. I wanted to flag this for the next round of reviews